File: src/main/java/com/vc/hard/L773.scala
package com.vc.hard

object L773 {
  def slidingPuzzle(board: Array[Array[Int]]): Int = {
    /**
            [
                [1,2,3],
                [4,0,5]
            ]
      */
    def generateState(): (String, Int) = {
      var start = ""
      board.indices.foreach(i => {
        board(i).indices.foreach(j => {
          start += board(i)(j)
        })
      })
      (start, start.indexOf('0'))
    }
    val target = "123450"
    val dirs = Array(
      Array(1, 3),
      Array(0, 2, 4),
      Array(1, 5),
      Array(0, 4),
      Array(1, 3, 5),
      Array(4, 2)
    )
    var res = 0
    import scala.collection.mutable
    val q = new mutable.Queue[(String, Int)]()
    val visited = new mutable.HashSet[String]()
    val start = generateState()
    q += start
    while(q.nonEmpty) {
      var size = q.size
      while(size > 0) {
        val e = q.dequeue()
        if(e._1 == target) return res
        if(!visited.contains(e._1)) {
          visited += e._1
          val oldIndex = e._2
          dirs(oldIndex).foreach(newIndex => {
            val c = e._1.toCharArray
            c(oldIndex) = c(newIndex)
            c(newIndex) = '0'
            q += ((c.mkString(""), newIndex))
          })
        }
        size -= 1
      }
      res += 1
    }
    -1
  }
}

File: src/main/java/com/vc/hard/L773.java
package com.vc.hard;

import java.util.*;

class L773 {

    static class Entry {
        String state;
        int emptySlot;

        Entry(String state, int emptySlot) {
            this.state = state;
            this.emptySlot = emptySlot;
        }

        @Override
        public String toString() {
            return ""+state+" "+emptySlot;
        }
    }

    private String target = "123450";
    private int[][] dirs = {
            {1, 3},
            {0, 2, 4},
            {1, 5},
            {0, 4},
            {1, 3, 5},
            {2, 4}
    };

    public int slidingPuzzle(int[][] board) {
        StringBuilder start = new StringBuilder();
        for(int i = 0; i < board.length; i++) {
            for(int j = 0; j < board[0].length; j++) {
                start.append(board[i][j]);
            }
        }

        Queue<Entry> q = new LinkedList<>();
        q.add(new Entry(start.toString(), start.toString().indexOf('0')));

        HashSet<String> visited = new HashSet<>();
        int level = 0;
        while(!q.isEmpty()) {
            int size = q.size();
            for(int i = 0; i < size; i++) {
                Entry entry = q.poll();
                //System.out.println(entry);

                if(entry.state.equals(target)) return level;
                if(visited.contains(entry.state+" "+entry.emptySlot)) continue;

                visited.add(entry.state+" "+entry.emptySlot);
                int oldIndex = entry.emptySlot;
                for(int newIndex: dirs[oldIndex]) {
                    char[] c = entry.state.toCharArray();
                    c[oldIndex] = c[newIndex];
                    c[newIndex] = '0';
                    String newState = new String(c);
                    q.offer(new Entry(newState, newIndex));
                }

            }
            level++;
        }
        return -1;
    }
}
